core/src/main/java/org/apache/iceberg/io/CDCWriter.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.PositionDelete;
+import org.apache.iceberg.util.StructLikeMap;
+import org.apache.iceberg.util.StructProjection;
+
+public class CDCWriter<T> extends BaseDeltaWriter<T> {
+
+  private final FanoutDataWriter<T> dataWriter;
+  private final PartitionAwareFileWriter<T, DeleteWriteResult> equalityDeleteWriter;
+  private final PartitionAwareFileWriter<PositionDelete<T>, DeleteWriteResult> positionDeleteWriter;
+  private final StructProjection keyProjection;
+  private final Map<StructLike, PartitionAwarePathOffset> insertedRows;
+  private final PositionDelete<T> positionDelete;
+  private final Function<T, StructLike> toStructLike;
+
+  public CDCWriter(FanoutDataWriter<T> dataWriter,
+                   PartitionAwareFileWriter<T, DeleteWriteResult> equalityDeleteWriter,
+                   PartitionAwareFileWriter<PositionDelete<T>, DeleteWriteResult> positionDeleteWriter,
+                   Schema schema, Schema deleteSchema, Function<T, StructLike> toStructLike) {
+    this.dataWriter = dataWriter;
+    this.equalityDeleteWriter = equalityDeleteWriter;
+    this.positionDeleteWriter = positionDeleteWriter;
+    this.positionDelete = new PositionDelete<>();
+    this.keyProjection = StructProjection.create(schema, deleteSchema);
+    this.insertedRows = StructLikeMap.create(deleteSchema.asStruct());
+    this.toStructLike = toStructLike;
+  }
+
+  @Override
+  public void insert(T row, PartitionSpec spec, StructLike partition) throws IOException {
+    CharSequence currentPath = dataWriter.currentPath(spec, partition);
```

Comment:
should we merge the two calls into one method? 
```
FilePathPosition pathPosition = dataWriter.filePathPosition(spec, partition);
```

Reply:
Since this is something that we will have to do for all insert rows, I was trying to avoid generating an extra object to reduce the memory consumption.